bears/c_languages/codeclone_detection/ClangComplexityBear.py

```diff
@@ -0,0 +1,91 @@
+from clang.cindex import Index, CursorKind
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.results.Result import Result
+from coalib.results.SourceRange import SourceRange
+from bears.c_languages.ClangBear import clang_available
+
+
+class ClangComplexityBear(LocalBear):
+    """
+    Calculates cyclomatic complexity of each function and displays it to the
+    user.
+    """
+    check_prerequisites = classmethod(clang_available)
+    decisive_cursor_kinds = {
+        CursorKind.IF_STMT, CursorKind.WHILE_STMT, CursorKind.FOR_STMT,
+        CursorKind.DEFAULT_STMT, CursorKind.CASE_STMT}
+
+    def function_key_points(self, cursor, top_function_level=False):
+        """
+        Calculates number of function's decision points and exit points.
+
+        :param top_function_level: Whether cursor is in the top level of
+                                   the function.
+        """
+        decisions, exits = 0, 0
+
+        for child in cursor.get_children():
+            if child.kind in self.decisive_cursor_kinds:
+                decisions += 1
+            elif child.kind == CursorKind.RETURN_STMT:
+                exits += 1
+                if top_function_level:
+                    # There is no point to move forward, so just return.
+                    return decisions, exits
+            child_decisions, child_exits = self.function_key_points(child)
+            decisions += child_decisions
+            exits += child_exits
+
+        if top_function_level:
+            # If return statement was avoided.
+            exits += 1
```

Comment:
I did it. Probably, not very good, but can't find a better way. :( Any
proposals? :)

12:42, пт, 25.03.2016 Lasse Schuirmann <notifications@github.com>:

> In bears/c_languages/codeclone_detection/ClangComplexityBear.py
> <https://github.com/coala-analyzer/coala-bears/pull/278#discussion_r57434877>
> :
>
> > +
> > +        for child in cursor.get_children():
> > +            if child.kind in self.decisive_cursor_kinds:
> > +                decisions += 1
> > +            elif child.kind == CursorKind.RETURN_STMT:
> > +                exits += 1
> > +                if top_function_level:
> > +                    # There is no point to move forward, so just return.
> > +                    return decisions, exits
> > +            child_decisions, child_exits = self.function_key_points(child)
> > +            decisions += child_decisions
> > +            exits += child_exits
> > +
> > +        if top_function_level:
> > +            # If return statement was avoided.
> > +            exits += 1
>
> ah right, maybe clarify in the comment?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/coala-analyzer/coala-bears/pull/278/files/324d45790841e0aa073017753ce98dac6241b681#r57434877>
>